lib/datadog/tracing/client_ip.rb

```diff
@@ -0,0 +1,191 @@
+    module ClientIp
+      DEFAULT_IP_HEADERS_NAMES = %w[
+        x-forwarded-for
+        x-real-ip
+        x-client-ip
+        x-forwarded
+        x-cluster-client-ip
+        forwarded-for
+        forwarded
+        via
+        true-client-ip
+      ].freeze
+
+      # A collection of headers.
+      class HeaderCollection
+        # Gets a single value of the header with the given name, case insensitive.
+        #
+        # @param [String] header_name Name of the header to get the value of.
+        # @returns [String, nil] A single value of the header, or nil if the header with
+        #   the given name is missing from the collection.
+        def get(header_name)
+          nil
+        end
+
+        def self.from_hash(hash)
+          HashHeaderCollection.new(hash)
+        end
+      end
+
+      # Sets the `http.client_ip` tag on the given span.
+      #
+      # This function respects the user's settings: if they disable the client IP tagging,
+      # or provide a different IP header name.
+      #
+      # @param [Span] span The span that's associated with the request.
+      # @param [HeaderCollection, #get, nil] headers A collection with the request headers.
+      # @param [String, nil] remote_ip The remote IP the request associated with the span is sent to.
+      def self.set_client_ip_tag(span, headers, remote_ip)
+        return if configuration.disabled
+
+        ip = client_address_from_request(headers, remote_ip)
+        if !configuration.header_name && ip.nil?
+          header_names = ip_headers(headers).keys
+          span.set_tag(TAG_MULTIPLE_IP_HEADERS, header_names.join(',')) unless header_names.empty?
+          return
+        end
+
+        unless valid_ip?(ip)
+          ip = extract_ip_from_full_address(ip)
+          return unless valid_ip?(ip)
+        end
+        ip = strip_zone_specifier(ip) if valid_ipv6?(ip)
+
+        span.set_tag(Tracing::Metadata::Ext::HTTP::TAG_CLIENT_IP, ip)
+      end
+
+      TAG_MULTIPLE_IP_HEADERS = '_dd.multiple-ip-headers'.freeze
+
+      # A header collection implementation that looks up headers in a Hash.
+      class HashHeaderCollection < HeaderCollection
+        def initialize(hash)
+          super()
+          @hash = hash.transform_keys(&:downcase)
+        end
+
+        def get(header_name)
+          @hash[header_name.downcase]
+        end
+      end
+
+      def self.ip_headers(headers)
+        return {} unless headers
+
+        {}.tap do |result|
+          DEFAULT_IP_HEADERS_NAMES.each do |name|
+            value = headers.get(name)
+            next if value.nil?
+
+            result[name] = value
+          end
+        end
+      end
+
+      def self.client_address_from_request(headers, remote_ip)
+        return headers.get(configuration.header_name) if configuration.header_name && headers
+
+        ip_values_from_headers = ip_headers(headers).values
+        case ip_values_from_headers.size
```

Comment:
```suggestion

        case ip_values_from_headers.size
```

Method return value of this method comes from the `case` return value, so I'd suggest to space this to conform with the Ruby style guide where implicit returns live on their own "line".

Reply:
Good point, done